kroxylicious-kms-provider-aws-kms/src/main/java/io/kroxylicious/kms/provider/aws/kms/credentials/Ec2MetadataCredentialsProvider.java

```diff
@@ -0,0 +1,318 @@
+            var thread = new Thread(r, Ec2MetadataCredentialsProvider.class.getName() + "thread");
+            thread.setDaemon(true);
+            return thread;
+        });
+        this.client = createClient();
+    }
+
+    private HttpClient createClient() {
+        var builder = HttpClient.newBuilder();
+        return builder
+                .followRedirects(HttpClient.Redirect.NORMAL)
+                .connectTimeout(HTTP_CONNECT_TIMEOUT)
+                .build();
+    }
+
+    @NonNull
+    @Override
+    public CompletionStage<SecurityCredentials> getCredentials() {
+        var newCredFuture = new CompletableFuture<SecurityCredentials>();
+        var witness = current.compareAndExchange(null, newCredFuture);
+        if (witness == null) {
+            // there's no current credential, let's create one
+            executorService.execute(() -> refreshCredential(newCredFuture));
+            return newCredFuture.minimalCompletionStage();
+        }
+        else if (isExpired(witness) || witness.isCompletedExceptionally()) {
+            // current credential is expired, or it has been completed exceptionally.
+            // throw it away and generate a new one.
+            // we don't normally expect to follow the expired path as the preemptive refresh ought to have
+            // caused its refresh before its expiration.
+            current.compareAndSet(witness, null);
+            return getCredentials();
+        }
+
+        return witness.minimalCompletionStage();
+    }
+
+    private void scheduleCredentialRefresh(long delay) {
+        LOGGER.debug("Scheduling refresh of AWS credentials in {}ms", delay);
+
+        var refreshedCredFuture = new CompletableFuture<SecurityCredentials>();
+        executorService.schedule(() -> {
+            refreshCredential(refreshedCredFuture);
+            refreshedCredFuture.thenApply(sc -> {
+                var previous = current.getAndSet(refreshedCredFuture);
+                // the previous future have been already complete, but for safety, complete it anyway.
+                Optional.ofNullable(previous).ifPresent(f -> f.complete(sc));
+                return null;
+            });
+        }, delay, TimeUnit.MILLISECONDS);
+    }
+
+    private boolean isExpired(CompletableFuture<SecurityCredentials> witness) {
+        if (witness.isDone() && !witness.isCompletedExceptionally()) {
+            try {
+                return Optional.ofNullable(witness.getNow(null))
+                        .map(SecurityCredentials::expiration)
+                        .map(exp -> systemClock.instant().isAfter(exp))
+                        .orElse(false);
+            }
+            catch (CancellationException | CompletionException e) {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private void refreshCredential(CompletableFuture<SecurityCredentials> future) {
```

Comment:
non-blocking: the JDK client has some [edge cases](https://bugs.openjdk.org/browse/JDK-8258397) where it can hang, we might want to also schedule a timeout job on the single-thread-executor to redundantly complete the future with timeout as an extra safety measure.